### DISCUSSION TITLE
Cloudberry Database Roadmap 2024

### DISCUSSION THREAD

**Discussion by tuhaihe** [apache/cloudberry#369] [Announcements]
> [!note]
> Hi everyone, since Cloudberry has joined the ASF Incubator, the new roadmap for Cloudberry is now up for discussion. We welcome your feedback: https://lists.apache.org/thread/wo1lsly8h7q8c4sxrwpw73ylbnzrcwrv.

---

---

This roadmap outlines the key milestones and focuses on the Cloudberry Database project in 2024. It aims to enhance functionality to strengthen further the Cloudberry Database's position as the next-gen unified open-source database for Analytics and AI.

> [!Note]
> Some plans may change depending on how the Cloudberry Database evolves in reality.

## Features and Enhancements

- [x] Implement federated queries across multiple Cloudberry Database clusters. (#320 in review)
- [x] Optimize query performance by leveraging materialized views.
- [ ] Refactor the dispatch logic for improved efficiency.
- [x] Enhance the access method interface. #391
- [x] Support directory table to manage files. #390 
- [x] Enable deployment on a single node. #206 
- [x] Support hybrid Row-Column storage for Cloudberry Database. (to be open sourced)
- [ ] Support vectorization to optimize the query performance. (to be open sourced)

## Kernel Update

- [ ] Upgrade Cloudberry Database kernel from PostgreSQL 14.4 to PostgreSQL 15.x stable release.

## Streaming

- [ ] Support streaming data loading.
- [x] Support the materialized views refresh mechanism, triggered by manual or automatic tasks, to enable near real-time analytics. (WIP)
- [x] Support Dynamic Table. #725 

## Lakehouse

- [ ] Integration with Apache Iceberg
- [ ] Integration with Apache Hudi
- [ ] Integration with Delta Lake, Hadoop

## AI/ML

- [x] MADlib for Cloudberry Database: [@cloudberrydb/madlib](https://github.com/cloudberrydb/madlib)
- [ ] Support the distributed Python compute engine via Ray for AI/ML workloads.

## Utilities, Tools, and Ecosystem

- [x] PostGIS for Cloudberry Database: [@cloudberrydb/postgis](https://github.com/cloudberrydb/postgis)
- [x] PXF for Cloudberry Database: [@cloudberrydb/pxf](https://github.com/cloudberrydb/pxf)
- [x] PL/R for Cloudberry Database: [@cloudberrydb/plr](https://github.com/cloudberrydb/plr)
- [x] PL/Java for Cloudberry Database: [@cloudberrydb/pljava](https://github.com/cloudberrydb/pljava)
- [ ] JDBC/ODBC for Cloudberry Database
- [ ] Container Service for Cloudberry Database
- [ ] K8S deployment support for Cloudberry Database
- [ ] Cloudberry Database Manager, one graphical tool with the deployment, SQL editor and monitoring dashboard, etc. (to be open sourced)
- [ ] Migration tool to help users migrate from Greenplum to Cloudberry Database.

## Release

We plan to release the Cloudberry Database at the following pace:

|      | v1.5.x | v1.6.x    | v1.7.x    |
| ---- | ------ | --------- | --------- |
| Jan  | ~1.5.0~  |           |           |
| Feb  | ~1.5.1~ / [1.5.0](https://github.com/cloudberrydb/cloudberrydb/releases/tag/1.5.0)  | | |
| Mar  | ~1.5.2~ / [1.5.1](https://github.com/cloudberrydb/cloudberrydb/releases/tag/1.5.1)  | | |
| Apr  |  [1.5.2](https://github.com/cloudberrydb/cloudberrydb/releases/tag/1.5.2) | ~1.6.0 rc1~ | |
| May  |  [1.5.3](https://github.com/cloudberrydb/cloudberrydb/releases/tag/1.5.3)<sub>pre-release</sub>      | ~1.6.0 rc2~ |           |
| Jun  |        | ~1.6.0~     |           |
| Jul  |  [1.5.4](https://github.com/cloudberrydb/cloudberrydb/releases/tag/1.5.4)      | ~1.6.1~   |           |
| Aug  |        | ~1.6.2~     |           |
| Sept |        |           | ~1.7.0 rc1~ |
| Oct  |        |           | ~1.7.0 rc2~ |
| Nov  |        |           | ~1.7.0~     |
| Dec  |        |           | ~1.7.1~     |

In addition, we will support more CPUs and operating systems for the release packages, including Linux for x86_64/aarch64, and KylinOS for x86_64/aarch64.

## Ask for feedbacks

We want to invite you to join us in shaping the future of Cloudberry Database. We'd love to hear your thoughts and ideas, so please don't hesitate to share them with us. We're committed to making our platform as open, user-friendly, and innovative as possible, and your feedback will help us achieve that goal. Thank you for your support, and we can't wait to hear from you!

---

[![Slack](https://img.shields.io/badge/Slack-6a32c9)](https://communityinviter.com/apps/cloudberrydb/welcome) [![Twitter](https://img.shields.io/twitter/follow/cloudberrydb)](https://twitter.com/cloudberrydb) [![WeChat](https://img.shields.io/badge/WeChat-eebc46)](https://cloudberrydb.org/community/wechat) [![Website](https://img.shields.io/badge/Website-bbec46)](https://cloudberrydb.org) [![Youtube](https://img.shields.io/badge/Youtube-gebc46)](https://youtube.com/@cloudberrydb) [![Discussions](https://img.shields.io/badge/Forum-gebc46)](https://github.com/orgs/cloudberrydb/discussions)

<sub>Want to try it out quickly? Follow our [Bootcamp program](https://github.com/cloudberrydb/bootcamp) to get started with the Cloudberry Database Sandbox.</sub>

**Comment by francksidi-edb** [comment]
Hi can you share what are the changes regarding compute/storage separation in the architecture and how we can dynamically increase/decrease number of segments ?

**Reply by tuhaihe** [reply]
For open-source Cloudberry, it is still in MPP architecture. The Cloudberry vendor, HashData, has implemented compute/storage separation in its commercial product. 

I am wondering where you are seeing the information. We can update it to avoid confusing anyone.

**Reply by tuhaihe** [reply]
We now support `gpshrink` and `gpexpand` to scale out/in segments.

for `gpshrink`, you can check the PR #393

**Comment by pjvldb** [comment]
Iceberg, PostGIS & GUI are all very smart moves.

**Reply by tuhaihe** [reply]
Thanks @pjvldb ❤️

**Comment by robozor** [comment]
Hello everyone,

I'm currently working with the Cloudberry Database and I'm interested in using the pgvector extension for vector similarity searches. Before proceeding, I wanted to ask if anyone has experience with installing or using pgvector within Cloudberry Database.

Is it compatible with Cloudberry, given its PostgreSQL 14.4 kernel? Are there any specific steps or considerations I should be aware of when setting it up?

Any insights or guidance would be greatly appreciated.

Thank you!

Josef Szylar

**Reply by francksidi-edb** [reply]
Hi, I successfully installed pgvector extension 0.7 on Cloudberry on a 1 node system.

**Reply by roseduan** [reply]
We have already integrated pgvector extension into Cloudberry.
https://github.com/cloudberrydb/pgvector

**Comment by reshke** [comment]
Is greenplum6->cloudberry migration native support in this roadmap? 
That is, upgrading existing greenplum6 deployment to cloudberry via pg_upgrade (gp_upgrade)

**Reply by tuhaihe** [reply]
yes, thanks for your reminder. will open source one business migration tool to help users migrate from GP to Cloudberry. I can udpate this to the roadmap later.

**Reply by tuhaihe** [reply]
Oh, you mean the native migration support, right? 
Any thoughts? @jianlirong @my-ship-it

**Reply by reshke** [reply]
> Oh, you mean the native migration support, right? Any thoughts? @jianlirong @my-ship-it

I mean something similar to pg_upgrade.

**Reply by jianlirong** [reply]
It seems like we need to do some investigations on the differences in system catalogs and data table formats between GP 6 and Cloudberry. Could you help us with this, @reshke ? Once we have a clear understanding of the differences, we can discuss the implementation options in more detail.

**Reply by reshke** [reply]
> It seems like we need to do some investigations on the differences in system catalogs and data table formats between GP 6 and Cloudberry. Could you help us with this, @reshke ? Once we have a clear understanding of the differences, we can discuss the implementation options in more detail.

Ok, i will do my best
as a starting point i propose https://github.com/cloudberrydb/cloudberrydb/pull/599

**Comment by techkuz** [comment]
does ```airflow.providers.postgres.operators.postgres import PostgresOperator``` support cloudberry?

**Reply by my-ship-it** [reply]
> does `airflow.providers.postgres.operators.postgres import PostgresOperator` support cloudberry?

Cloudberry keeps interface compatibility with postgres.  So integration with airflow seems not hard. If you found any issues when integrating with airflow, please file issues here: https://github.com/apache/cloudberry/issues

**Comment by Xedonedron** [comment]
`Cloudberry Database Manager`
Does this reference from Greenplum Command Center with more powerful feature like querying or administration things?

**Reply by my-ship-it** [reply]
> `Cloudberry Database Manager` Does this reference from Greenplum Command Center with more powerful feature like querying or administration things?

Yes, it is.

**Comment by adar604** [comment]
Cart

**Comment by tuhaihe** [comment]
Hi everyone, since Cloudberry has joined the ASF Incubator, the new roadmap for Cloudberry is now up for discussion. We welcome your feedback: https://lists.apache.org/thread/wo1lsly8h7q8c4sxrwpw73ylbnzrcwrv.

**Comment by sobolevna** [comment]
Hello! What type of integration with Iceberg do you mean? Native, PXF, some other methods? 
And are you planning to integrate with Paimon tables?

**Reply by jianlirong** [reply]
What do you mean by "Native"? We are working on a FDW named datalake_fdw, which will support the functionalities of access to Iceberg tables(including read and write). That FDW will be open sourced soon. That's the initial plan of integration with Iceberg in our head. On the other hand, I think another group from the OSS community is working on upgrading PXF, and integration with Iceberg maybe one of the upgrade jobs, I guess.

Any idea about integration with Iceberg or Paimon tables is welcome.

**Reply by yjhjstz** [reply]
We can refer to paper https://dl.acm.org/doi/pdf/10.1145/3722212.3724459 from databricks, make Iceberg as first class table like heap table that called "Native".

**Reply by yjhjstz** [reply]
I found this https://www.crunchydata.com/blog/crunchy-data-warehouse-postgres-with-iceberg-for-high-performance-analytics to prove my standpoint.

**Comment by xanadu-dev** [comment]
Is there any update on the schedule for apache-cloudberry 2.0 full release ?

**Reply by tuhaihe** [reply]
Hi @xanadu-dev, we now have two rounds for 2.0—RC1 and RC2—and we've fixed some license issues. I think we'll have an RC3 round soon (this week?).

As you may know, Cloudberry, as an incubator project, will go through a two-phase vote (for the Cloudberry PPMC vote, then for the Apache Incubator vote). Regarding the schedule, I think we can not set a fixed time now, but we will speed up the process ASAP.

Related links:
- RC1 VOTE: https://lists.apache.org/thread/o7ysvxjw2dxy7bvc6o4x25sy9tzv1975
- RC2 VOTE: https://lists.apache.org/thread/9mgmq51650j7o767rss9t9ynh3bo1yvx
